include/unifex/upon_done.hpp

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/async_trace.hpp>
+#include <unifex/bind_back.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/std_concepts.hpp>
+#include <unifex/tag_invoke.hpp>
+#include <unifex/type_list.hpp>
+#include <unifex/type_traits.hpp>
+
+#include <functional>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+namespace _upon_done {
+
+template <typename Receiver, std::invocable Func>
+struct _receiver {
+  struct type;
+};
+template <typename Receiver, std::invocable Func>
+using receiver_t = typename _receiver<Receiver, Func>::type;
+
+template <typename Receiver, std::invocable Func>
+struct _receiver<Receiver, Func>::type {
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+  template <typename... Values>
+  void set_value(Values&&... values) && noexcept {
+    unifex::set_value(std::move(receiver_), std::forward<Values>(values)...);
```

Comment:
Rather than using `std::move` and `std::forward` (which instantiate templates), we generally use `static_cast` directly to save compile time. And to be honest, we tend to use C-style casts for that (but don't tell Bjarne ;-).